src/query/functions/temporal/base.go

```diff
@@ -0,0 +1,505 @@
+func (c *baseNode) processCompletedBlocks(processRequests []processRequest, queryStartBounds, queryEndBounds block.Bounds, maxBlocks int) error {
+	processedKeys := make([]time.Time, len(processRequests))
+	for i, req := range processRequests {
+		if err := c.processSingleRequest(req); err != nil {
+			return err
+		}
+
+		processedKeys[i] = req.bounds.Start
+	}
+
+	// Mark all blocks as processed
+	c.cache.markProcessed(processedKeys)
+	// Sweep to free blocks from cache with no dependencies
+	c.sweep(c.cache.processed(), queryStartBounds, queryEndBounds, maxBlocks)
+	return nil
+}
+
+func (c *baseNode) processSingleRequest(request processRequest) error {
+	aggDuration := c.op.duration
+	seriesIter, err := request.blk.SeriesIter()
+	if err != nil {
+		return err
+	}
+
+	depIters := make([]block.SeriesIter, len(request.deps))
+	for i, blk := range request.deps {
+		iter, err := blk.SeriesIter()
+		if err != nil {
+			return err
+		}
+		depIters[i] = iter
+	}
+
+	bounds := seriesIter.Meta().Bounds
+	steps := int((aggDuration + bounds.Duration) / bounds.StepSize)
+	values := make([]float64, 0, steps)
+
+	seriesMeta := seriesIter.SeriesMeta()
+	resultSeriesMeta := make([]block.SeriesMeta, len(seriesMeta))
+	for i, m := range seriesMeta {
+		tags := m.Tags.WithoutName()
+		resultSeriesMeta[i].Tags = tags
+		resultSeriesMeta[i].Name = tags.ID()
+	}
+
+	builder, err := c.controller.BlockBuilder(seriesIter.Meta(), resultSeriesMeta)
+	if err != nil {
+		return err
+	}
+
+	if err := builder.AddCols(bounds.Steps()); err != nil {
+		return err
+	}
+
+	for seriesIter.Next() {
+		values = values[0:0]
+		for i, iter := range depIters {
+			if !iter.Next() {
+				return fmt.Errorf("incorrect number of series for block: %d", i)
+			}
+
+			s, err := iter.Current()
+			if err != nil {
+				return err
+			}
+
+			values = append(values, s.Values()...)
+		}
+
+		desiredLength := int(aggDuration / bounds.StepSize)
+		series, err := seriesIter.Current()
+		if err != nil {
+			return err
+		}
+
+		for i := 0; i < series.Len(); i++ {
+			val := series.ValueAtStep(i)
+			values = append(values, val)
+			newVal := math.NaN()
+			if desiredLength <= len(values) {
+				values = values[len(values)-desiredLength:]
+				newVal = c.processor.Process(values)
+			}
+
+			builder.AppendValue(i, newVal)
+
+		}
+	}
+
+	nextBlock := builder.Build()
+	defer nextBlock.Close()
+	return c.controller.Process(nextBlock)
+}
+
+func (c *baseNode) sweep(processedKeys []bool, queryStartBounds block.Bounds, queryEndBounds block.Bounds, maxBlocks int) {
+	prevProcessed := 0
+	numBlocks := len(processedKeys) - 1
+	for i := numBlocks; i >= 0; i-- {
+		processed := processedKeys[i]
+		if !processed {
+			prevProcessed = 0
+			continue
+		}
+
+		dependantBlocks := maxBlocks
+		if dependantBlocks > numBlocks-i-1 {
+			dependantBlocks = numBlocks - i - 1
+		}
+
+		if prevProcessed >= dependantBlocks {
+			c.cache.remove(i)
```

Comment:
This seems like it would cause issues considering it's removing at an index which changes independently of here

i.e. if we're deleting at index 0,1,2 from [0,1,2], wouldn't it go something like [0,1,2]->[1,2]->[1]->out of index?

Reply:
added a test for that and it seems to work :)